libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -510,3 +630,246 @@ void ContractLevelChecker::checkBaseABICompatibility(ContractDefinition const& _
 		);
 
 }
+
+void ContractLevelChecker::checkAmbiguousOverrides(ContractDefinition const& _contract) const
+{
+	vector<FunctionDefinition const*> contractFuncs = _contract.definedFunctions();
+
+	auto const resolvedBases = resolveBaseContracts(_contract);
+
+	FunctionSet collectedFunctions;
+
+	// Collect the most inherited functions of the direct base contracts
+	for (ContractDefinition const* contract: resolveBaseContracts(_contract))
+	{
+		FunctionSet baseSet = convertContainer<FunctionSet>(contract->definedFunctions());
+
+		for (FunctionDefinition const* function: getBaseFunctions(contract))
+		{
+			if (baseSet.cend() == baseSet.find(function))
+				baseSet.insert(function);
+		}
+
+		if (baseSet.empty())
+			continue;
+
+		// Collect the first of every override-set of functions (they are sorted
+		// by most-inherited function)
+		for (auto it = baseSet.cbegin(); it != baseSet.cend(); it = baseSet.upper_bound(*it))
+		{
+			if (!(*it)->isOverridable())
+				continue;
+
+			collectedFunctions.insert(*it);
+		}
+	}
+
+	// Check the sets of the most-inherited functions
+	for (auto it = collectedFunctions.cbegin(); it != collectedFunctions.cend(); it = collectedFunctions.upper_bound(*it))
+	{
+		auto [begin,end] = collectedFunctions.equal_range(*it);
+
+		// Only one function
+		if (next(begin) == end)
+			continue;
+
+		// Not an overridable function
+		if (!(*it)->isOverridable())
+			continue;
+
+		// Function has been explicitly overridden
+		if (contains_if(
+				contractFuncs,
+				[&] (FunctionDefinition const* _f) {
+					return hasEqualNameAndParameters(*_f, **it);
+				}
+			)
+		)
+			continue;
+
+		set<FunctionDefinition const*> ambiguousFunctions;
+		SecondarySourceLocation ssl;
+
+		for (;begin != end; begin++)
+		{
+			ambiguousFunctions.insert(*begin);
+			ssl.append("Definition here: ", (*begin)->location());
+		}
+
+		// Make sure the functions are not from the same base contract
+		if (ambiguousFunctions.size() == 1)
+			continue;
+
+		m_errorReporter.typeError(
+			_contract.location(),
+			ssl,
+			"Derived contract must override function " +
+			(*it)->name() +
+			". Function with the same name and parameter types defined in two or more base classes."
+		);
+	}
+}
+
+set<ContractDefinition const*> ContractLevelChecker::resolveOverrideList(OverrideSpecifier const& _overrides) const
```

Comment:
Sets of pointers are rather dangerous for determinism. If we check for duplicates by AST ID, then maybe you can change this to a vector?

Reply:
A set with this type is also used in another place where we rely on the unique inserts as well. For the sets I added a special less compare function for contracts that compares the id()'s now instead.